### PR TITLE
fix a couple bugs with circom source functions that return vector values

### PR DESCRIPTION
**Pull request by tim-hoffman**
- add more tests
- other various code improvements

### CODE REVIEW

File: compiler/src/intermediate_representation/call_bucket.rs [hunk 1]
```diff
@@ -107,7 +109,7 @@ impl WriteLLVMIR for CallBucket {
         Self::manage_debug_loc_from_curr(producer, self);
 
         // Check arena_size==0 which indicates arguments should not be placed into arena
-        let arena_size = self.arena_size;
+        let arena_size = u32::try_from(self.arena_size).expect("overflow");
```
Comment (0xddom) [inline compiler/src/intermediate_representation/call_bucket.rs:112]: Why do we need to cast to 32bits here?

**Review by 0xddom — APPROVED** [review]
Just a nit about a cast, other than that LGTM

File: code_producers/src/llvm_elements/functions.rs [hunk 2]
```diff
@@ -77,16 +77,16 @@ impl<'a> FunctionCtx<'a> {
...
         index: IntValue<'a>,
-    ) -> AnyValueEnum<'a> {
+    ) -> PointerValue<'a> {
         create_gep(producer, self.arena, &[index])
     }
 
-    fn get_variable_array(&self, _producer: &dyn LLVMIRProducer<'a>) -> AnyValueEnum<'a> {
-        self.arena.into()
+    fn get_variable_array(&self, _producer: &dyn LLVMIRProducer<'a>) -> PointerValue<'a> {
+        self.arena
```
Comment (iangneal) [inline code_producers/src/llvm_elements/functions.rs:89]: This is unrelated to this PR, but could we also output the arena size as part of the JSON summary?

**Review by iangneal — APPROVED** [review]
Also looks good to me, just a random thought

File: compiler/src/intermediate_representation/call_bucket.rs (hunk 1 above)
Comment (tim-hoffman) [reply compiler/src/intermediate_representation/call_bucket.rs:112]: I'll move it later. The array_type function needs it.

File: code_producers/src/llvm_elements/functions.rs (hunk 2 above)
Comment (tim-hoffman) [reply code_producers/src/llvm_elements/functions.rs:89]: That one's a bit tricky because the arena size is not computed yet but I'll look into it.

**Pull request merged**